src/gt4py/next/iterator/transforms/global_tmps.py

```diff
@@ -19,11 +19,73 @@
     domain_utils,
     ir_makers as im,
 )
+from gt4py.next.iterator.ir_utils.domain_utils import SymbolicDomain
 from gt4py.next.iterator.transforms import cse, infer_domain, inline_lambdas
 from gt4py.next.iterator.type_system import inference as type_inference
 from gt4py.next.type_system import type_info, type_specifications as ts
 
 
+def _filter_setat_by_domain(stmt: itir.SetAt, domain: SymbolicDomain) -> itir.SetAt:
+    assert cpm.is_call_to(stmt.expr, "make_tuple")
+
+    new_targets = []
+    new_els = []
+    # we cannot use `el.annex.domain` since while it might be semantically equivalent, it does
+    # not necessarily be equal.
+    # can be reproduced in icon4py `dycore_tests/test_solve_nonhydro.py::test_nonhydro_predictor_step`
+    # compute_theta_rho_face_values_and_pressure_gradient_and_update_vn
+    assert isinstance(stmt.expr.annex.domain, tuple)
+    assert domain in stmt.expr.annex.domain
+    for i, (el, el_domain) in enumerate(zip(stmt.expr.args, stmt.expr.annex.domain)):
+        assert not isinstance(el_domain, tuple), "Nested tuples not supported."
+        assert isinstance(el_domain, SymbolicDomain)
+        if el_domain == domain:
+            new_targets.append(im.tuple_get(i, stmt.target))
+            new_els.append(el)
```

Comment:
Just two suggestions: 
- I would move the assertions about the inputs to the beginning of the function
- I would move the asserts per element to the call site outside this function, since I think it's more clear and also avoids asserting multiple times the same conditions.
```python 
            assert not isinstance(el_domain, tuple), "Nested tuples not supported."
            assert isinstance(el_domain, SymbolicDomain)
```

```suggestion
    assert cpm.is_call_to(stmt.expr, "make_tuple")
    assert isinstance(stmt.expr.annex.domain, tuple)
    assert domain in stmt.expr.annex.domain

    new_targets = []
    new_els = []
    # we cannot use `el.annex.domain` since while it might be semantically equivalent, it does
    # not necessarily be equal.
    # can be reproduced in icon4py `dycore_tests/test_solve_nonhydro.py::test_nonhydro_predictor_step`
    # compute_theta_rho_face_values_and_pressure_gradient_and_update_vn
    for i, (el, el_domain) in enumerate(zip(stmt.expr.args, stmt.expr.annex.domain)):
        if el_domain == domain:
            new_targets.append(im.tuple_get(i, stmt.target))
            new_els.append(el)
```